File: ScriptTester/src/main/scala/Utilities.scala
/**
 * Created by marcus on 12/17/14.
 */

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import org.luaj.vm2.LuaTable;

import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.io._;
import java.net.URLConnection;
import java.util.concurrent.ExecutionException;
import java.util.Date;
import java.nio.ByteBuffer;

import scala.collection.JavaConversions._
import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Await
import scala.concurrent.duration._

object Utilities {
    //val executor = if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.GINGERBREAD_MR1) {
    /*val executor = if (true) {
      new Executor  {
        def execute(r: Runnable) {
          new Thread(r).start();
        }
      }
    } else {
      //AsyncTask.THREAD_POOL_EXECUTOR
    }
    implicit val exec = ExecutionContext.fromExecutor(executor)*/
    var gson : Gson = null


    def getGson() : Gson = {
        if (gson == null)
            gson = new GsonBuilder().registerTypeAdapter(classOf[LuaTable], new LuaTableSerializer()).setPrettyPrinting().create()
        return gson
    }

    var id = 0
    def getID() : Int = this.synchronized { id += 1; id }

    def download(source : String) : String = {
        try {

            val f : Future[String] = Future[String] {
                DownloadSource(source)
            }
            //Should Probably make the app handle timeouts a bit better
            return Await.result(f,60 seconds).asInstanceOf[String]
        }
        catch {
            case e : ExecutionException => {
                error("Download: Error with Async Source Download!!!")
                error(e.getMessage())
                var err = ""
                for(x <- e.getStackTrace()) {
                    err += x.toString()+"\n"
                }
                error(err)
                return "Error Async!"
            }
            case e : InterruptedException => {
                error("Download: Interrupted Exception!!!")
                error(e.getMessage())
                return "Error Interrupted!"
            }
        }
    }


    def DownloadSource(source : String) : String = {
        var filename : String = ""
        var resultingPath : String = ""
        var sourceSite : URL = null
        if(!source.isEmpty())
        {
            debug("DownloadSource: "+source)
            try {
                error("DownloadSource: URL is: "+source)
                sourceSite = new URL(source)
                if (sourceSite == null) {
                  error("HUGE ERROR!!!")
                }
                var urlcon : URLConnection = sourceSite.openConnection()
                //Use custom user agent to prevent http 403 response
                //If custom user agent doesn't work use actual one
                //urlcon.setRequestProperty("User-Agent", "Mozilla/5.0 (X11; Linux x86_64; rv:41.0) Gecko/20100101 Firefox/41");
                urlcon.setRequestProperty("User-Agent", "Mangagaga");
                error("TYPE IS... "+urlcon.getContentType)       
                var extension : String = ""
                if (source.contains(".jpg")) {
                    extension = ".jpg"
                } else if (source.contains(".jpeg")) {
                    extension = ".jpeg"
                } else if (source.contains(".png")) {
                    extension = ".png"
                } else if (source.contains(".zip")) {
                    extension = ".zip"
                } else if (source.contains(".apk")) {
                    extension = ".apk"
                } else {
                    extension = ".html"
                }
                info("DONLSDFING: "+source)
                var extensionIndex : Int = source.lastIndexOf(extension)
                if (extensionIndex != -1)
                    filename = source.substring((source.lastIndexOf('/') + 1), extensionIndex + extension.length())
                else
                    filename = source.substring((source.lastIndexOf('/') + 1)) + extension

                var dest : String = "./Mangagaga/Cache/"

                if (filename.contains("?")) {
                    filename = filename.replace('?', '_')
                    debug("DownloadSource: Removed '?' and renamed file to: " + filename)
                }
                // Prepend an ID to the file name so different files do not conflict
                filename = getID().toString + "_" + filename

                resultingPath = dest + filename
                var file : File = new File(dest, filename)
                info("DownloadSource: resulting file: "+file.toString())
                try {
                    file.createNewFile()
                } catch {
                    case e : IOException => {
                        error("DownloadSource: Couldn't make file!!!")
                        error("DownloadSource: "+e.toString())
                    }
                }

                debug("DownloadSource: Making Image fos")
                var fos : FileOutputStream = new FileOutputStream(file)
                info("DownloadSource: FOS is made!")
                //Error on this line...
                var is : InputStream = urlcon.getInputStream()
                info("DownloadSource: IS is made!")
                var bis : BufferedInputStream = new BufferedInputStream(is)
                info("DownloadSource: BIS is made!")
                var buffer : ByteBuffer = ByteBuffer.allocate(1024 * 1024 * 16)
                info("DownloadSource: buffer is made!")

                var chunk : Int = bis.read()

                while (chunk != -1) {
                  buffer.put(chunk.asInstanceOf[Byte])
                  chunk = bis.read()
                }
                debug("DownloadSource: Writing Image")
                fos.write(buffer.array)
                fos.flush()
                fos.close()
            }
            catch {
                case e : MalformedURLException => {
                    error("DownloadSource: Error opening page!")
                    return "Error 1"
                }
                case e : IOException => {
                    error("DownloadSource: Error With Reader/Writer!")
                    error("DownloadSource: "+e.toString())
                    return "Error 2"
                }
            }
        }

        return resultingPath
    }

    @throws(classOf[IOException])
    def readFile(absolutePath : String) : String = {
        return readFile(new FileInputStream(absolutePath))
    }

    @throws(classOf[IOException])
    def readFile(input : InputStream) : String = {
        var reader : BufferedReader = new BufferedReader(new InputStreamReader(input))
        var sb : StringBuilder = new StringBuilder()
        var line : String = reader.readLine()
        while (line != null) {
            sb.append(line).append("\n")
            line = reader.readLine()
        }
        /*for (var line : String = reader.readLine(); line != null; line = reader.readLine())
            sb.append(line).append("\n");*/
        return sb.toString()
    }

    @throws(classOf[IOException])
    def copyStreams(in : InputStream, out : OutputStream) {
        var buffer : Array[Byte] = new Array[Byte](1024);
        var readBytes : Int = in.read(buffer)
        while(readBytes != -1) {
            out.write(buffer, 0, readBytes)
            readBytes = in.read(buffer)
        }
        return
    }

    def clearCache()
    {
        var cache : File = new File("./Mangagaga/Cache/")
        clearFolder(cache)
    }

    def deleteFolder(folder : File) {
      if (folder.exists()) {
        clearFolder(folder)
        folder.delete()
      }
    }

    def clearFolder(folder : File) {

        if(!folder.exists())
        {
            error("clear saved: Error clearing saved, directory doesn't exist")
        }else {
            for (child <- folder.list()) {
                var childFile : File = new File(folder, child)
                if (childFile.isDirectory())
                    clearFolder(childFile)
                childFile.delete()
            }
        }
    }

    def error(in : String) {
      System.out.println(in)
    }

    def info(in : String) {
      System.out.println(in)
    }

    def debug(in : String) {
      System.out.println(in)
    }
}

File: common/src/main/java/io/githup/limvot/mangagaga/Utilities.kt
package io.githup.limvot.mangagaga;

/**
 * Created by marcus on 12/17/14. (redone in Kotlin later by Nathan :D)
 */

import java.net.HttpURLConnection;
import java.net.CookieHandler;
import java.net.CookieManager;
import java.net.MalformedURLException;
import java.net.URL;
import java.io.*;
import java.util.Date;

import org.eclipse.jgit.api.Git
import org.eclipse.jgit.transport.UsernamePasswordCredentialsProvider

object Utilities : GenericLogger {

    init {
        CookieHandler.setDefault(CookieManager())
    }

    var id = 0
    fun getID() = synchronized(this) { id += 1; id }

    fun getModifiedTime(source : String) : Date {
        info("Url to grab modified time from $source")
        var modifiedTime : Long = 0
        try {
            val connection = URL(source).openConnection() as HttpURLConnection
            connection.setRequestMethod("HEAD")
            connection.connect()
            modifiedTime = connection.getLastModified()
            connection.disconnect()
        } catch(e : Exception) {
            info("getModifiedTime "+e.toString())
            modifiedTime = 0
        }
        if (modifiedTime.toInt() == 0)
            info("Could not get modified time, Nope couldn't get it")
        return Date(modifiedTime)
    }

    fun download(source : String) = downloadWithRequestHeadersAndReferrer(source, "").first
    fun downloadWithRequestHeadersAndReferrer(source : String, referer: String) : Pair<String, MutableMap<String,List<String>>> {
        // Removed a lot of try/catch/timeout stuff here.
        // We should add it back if we need it
        return DownloadSource(source, referer)
    }

    fun DownloadSource(source : String, referer: String) : Pair<String, MutableMap<String,List<String>>> {
        var resultingPath : String = ""
        var resultingHeaders : MutableMap<String,List<String>>? = null
        info("DownloadSource: $source")
        var urlcon : HttpURLConnection? = null
        try {
            info("DownloadSource: URL is: $source")
            urlcon = URL(source).openConnection() as HttpURLConnection
            urlcon.setRequestProperty("User-Agent", "Mozilla/5.0 (X11; Linux x86_64; rv:44.0) Gecko/20100101 Firefox/44.0");
            if (referer.isNotEmpty())
              urlcon.setRequestProperty("Referer", referer);
            //urlcon.setRequestProperty("User-Agent", "Mangagaga");
            info("user agent is ${urlcon.getRequestProperty("User-Agent")}")
            urlcon.setInstanceFollowRedirects(true)
            info("TYPE IS... ${urlcon.getContentType()}")       
            var filename : String = if (referer.isNotEmpty()) {
              info("Using referer instead: $referer")
              referer.substring((source.lastIndexOf('/') + 1))
            } else {
                source.substring((source.lastIndexOf('/') + 1))
            }

            filename = filename.replace('?', '_')
            info("filename: $filename")

            val dest = SettingsManager.mangagagaPath + "/Cache/"
            // Prepend an ID to the file name so different files do not conflict
            filename = "${getID()}_$filename"

            resultingPath = dest + filename
            val file = File(dest, filename)
            info("DownloadSource: resulting file: $file")
            file.createNewFile()
            
            resultingHeaders = urlcon.getHeaderFields()
            info("Request Headers: $resultingHeaders")
            info("here we go")
            info(urlcon.getResponseMessage())

            // RRRARRRARARGH you gave to do getErrorStream if it's returned an info
            // and getResponseMessage doesn't work if it returned an info
            val ist = if (urlcon.getResponseCode() in 200..299)
                urlcon.getInputStream()
            else
              urlcon.getErrorStream()
            file.writeBytes(ist.readBytes())
        }
        catch (e : MalformedURLException) {
            info("DownloadSource: Error opening page!")
            return Pair("Error 1 - malformed url", resultingHeaders!!)
        }
        catch (e : IOException) {
            info("DownloadSource: Error With Reader/Writer! (returning getResponseMessage as the string instead of the file path)")
            info("DownloadSource: $e")
            return Pair("Error 2 info with reader writer", resultingHeaders!!)
        }
        finally {
            urlcon?.disconnect()
        }

        return Pair(resultingPath, resultingHeaders!!)
    }

    fun clearCache() {
        val cache = File(SettingsManager.mangagagaPath + "/Cache/")
        clearFolder(cache)
    }

    fun deleteFolder(folder : File) {
      if (folder.exists()) {
        clearFolder(folder)
        folder.delete()
      }
    }

    fun clearFolder(folder : File) {
        for (child in folder.list()) {
            val childFile : File = File(folder, child)
            if (childFile.isDirectory())
                clearFolder(childFile)
            childFile.delete()
        }
    }

    private fun gitScriptHelper(f_script: File?, username: String, password: String, commit_msg: String) {
        val to = File(SettingsManager.mangagagaPath, "Scripts/")
        val git_temp = File(SettingsManager.mangagagaPath, "GitTemp/")
        Utilities.deleteFolder(git_temp)
        val result = Git.cloneRepository().setURI(SettingsManager.getGitURL())
                                          .setDirectory(git_temp)
                                          .call()

        if(f_script == null) {
            //gitToScripts()
            for (script in git_temp.listFiles().filter { it.isFile() }) {
                File(to, script.name).writeText(File(script.absolutePath).readText())
            }
        } else {
            //scriptToGit()
            val script = f_script
            File(git_temp, script.name).writeText(script.readText())
            result.add().addFilepattern(script.name).call()
            result.commit().setMessage(commit_msg).call()
            result.push().setCredentialsProvider(
                    UsernamePasswordCredentialsProvider(username, password)).call()
        }
    }
    fun gitToScripts() {
        gitScriptHelper(null,"","","")
    }

    fun scriptToGit(script: File, username: String, password: String, commit_msg: String) {
        gitScriptHelper(script, username, password, commit_msg)
    }
}
